autopilot/betweenness_centrality.go

```diff
@@ -0,0 +1,177 @@
+package autopilot
+
+// stack is a simple int stack to help with readability of Brandes'
+// betweenness centrality implementation below.
+type stack struct {
+	stack []int
+}
+
+func (s *stack) Push(v int) {
+	s.stack = append(s.stack, v)
+}
+
+func (s *stack) Top() int {
+	return s.stack[len(s.stack)-1]
+}
+
+func (s *stack) Pop() {
+	s.stack = s.stack[:len(s.stack)-1]
+}
+
+func (s *stack) Empty() bool {
+	return len(s.stack) == 0
+}
+
+// queue is a simple int queue to help with readability of Brandes'
+// betweenness centrality implementation below.
+type queue struct {
+	queue []int
+}
+
+func (q *queue) Push(v int) {
+	q.queue = append(q.queue, v)
+}
+
+func (q *queue) Front() int {
+	return q.queue[0]
+}
+
+func (q *queue) Pop() {
+	q.queue = q.queue[1:]
+}
+
+func (q *queue) Empty() bool {
+	return len(q.queue) == 0
+}
+
+// BetweennessCentrality is a NodeMetric that calculates node betweenness
+// centrality using Brandes' algorithm. Betweenness centrality for each node
+// is the number of shortest paths passing trough that node, not counting
+// shortest paths starting or ending at that node. This is a useful metric
+// to measure control of individual nodes over the whole network.
+type BetweennessCentrality struct {
+	centrality map[NodeID]float64
+}
+
+// NewBetweennessCentralityMetric creates a new BetweennessCentrality instance.
+func NewBetweennessCentralityMetric() *BetweennessCentrality {
+	return &BetweennessCentrality{}
+}
+
+// Name returns the name of the metric.
+func (bc *BetweennessCentrality) Name() string {
+	return "betweeness_centrality"
+}
+
+// betweennessCentrality is the core of Brandes' algorithm.
+// We first calculate the shortest paths from the start node s to all other
+// nodes with BFS, then update the betweenness centrality values by using
+// Brandes' dependency trick.
+// For detailed explanation please read:
+// https://www.cl.cam.ac.uk/teaching/1617/MLRD/handbook/brandes.html
```

Comment:
Thinking about this a bit more, this algorithm is fine for the case where we want to calculate the value for _every_ node. However, for the autopilot case, we only care about _our_ centrality metric, so perhaps we'll need to adopt a slightly modified algorithm which can be more efficient when we only care about a single node.

Reply:
There are two reasons for calculating the betweenness centrality for all nodes:

1) to get the centrality value for a single node, we have to calculate all shortest paths for every s, t pairs where s != t, this is required as the centrality of a single vertex is the sum(sigma(s,t,v) / sigma(s,t)) [for every s != t != v], where sigma(s,t) is the number of shortest paths between s,t and sigma(s,t,v) is the number of the shortest paths between s,t that pass trough v. As calculating all paths is the bulk of the computation, the rest of the algorithm is much less time consuming, and so we wouldn't spare much.

2) betweenness centrality for a single node doesn't really have much meaning as it depends on how connected the graph is not an absolute measure.